### ISSUE TITLE
behaviour of option+backspace differs between v3 and v4

### ISSUE THREAD

**Issue by ZuBB** [fish-shell/fish-shell#10926]
Steps to reproduce:
* Install fish v4b1. I got mine from [here](https://github.com/fish-shell/homebrew-fish-beta-4)
* Type in the terminal something like this `vim .config/git/.gituser`
* Press `option`+`backspace` shortcut

Expected behaviour: one word starting from the end gets deleted
Actual behavior: whole path gets deleted

I am not sure if you consider this a bug, but behavior has definitely been changed

Env
```
vv@my-precious ~ $ fish --version
fish, version 4.0b1
vv@my-precious ~ $ kitty --version
kitty 0.38.0 created by Kovid Goyal
vv@my-precious ~ $ uname -a
Darwin my-precious.local 23.6.0 Darwin Kernel Version 23.6.0: Fri Nov 15 15:13:15 PST 2024; root:xnu-10063.141.1.702.7~1/RELEASE_ARM64_T6000 arm64
vv@my-precious ~ $ echo $TERM
xterm-kitty
vv@my-precious ~ $ 
```

**Comment by krobelus** [comment]
yes that was intended (more or less). As a workaround you can use `bind alt-backspace backward-kill-word`.
The default for that is `ctrl-backspace`, which doesn't work on macOS unfortunately IIRC.

Previous discussions are

- https://github.com/fish-shell/fish-shell/pull/10766
- https://github.com/fish-shell/fish-shell/commit/6af96a81a8c6ff5e632d4dda7448f5f01d7a5d35#r147894640
- https://github.com/fish-shell/fish-shell/issues/10910#issuecomment-2548142551

**Comment by ypnos** [comment]
It's right there in the Release Notes as well: https://github.com/fish-shell/fish-shell/releases/tag/4.0b1

**Comment by ZuBB** [comment]
thanks @krobelus, I think I will go with a `bind` option

**Comment by krobelus** [comment]
> The default for that is ctrl-backspace

actually `ctrl-backspace` (which is mentioned in the changelog) does work on it iTer. Only Terminal.app can't disambiguate it.
Also tmux but there is ongoing work to fix that.

**Comment by chriskrycho** [comment]
Can you link to the PR or notes on why this changed, just so folks like me can understand? I come with the default assumption that there’s a good reason for changes like this, but I also feel mildly grumpy because this breaks muscle memory that I have been building up for a decade now, and I would very much prefer *not* to use `bind` to override the defaults, because that makes portability much more annoying. Is this aimed at matching the behavior of other shells (I note after testing that this matches what zsh does but *not* what bash does) or something else? Again, I’m very much open to hearing that there’s a strong motivation for it, *and* it’s the one thing that has felt frustrating since installing v4b1.

**Comment by krobelus** [comment]
> Can you link to the PR or notes on why this changed, just so folks like me can understand?

Ok, I guess the existing changelog notes do mention the alternative in `ctrl-backspace`
but don't elaborate.

> I come with the default assumption that there’s a good reason for changes like this,

I think part of the decision depends on what users expect on average. Boldly making the change is an easy/lazy way to get this information...

> but I also feel mildly grumpy because this breaks muscle memory that I have been building
> up for a decade now, and I would very much prefer not to use bind to override the defaults,
> because that makes portability much more annoying.

Yes, absolutely.

> Is this aimed at matching the behavior of other shells

Not really; we already deviate from readline in `ctrl-w`, so we want to complement our set of bindings.

> (I note after testing that this matches what zsh does but not what bash does)

Can't reproduce; zsh behaves the same as bash here, `alt-backspace` deletes only a word.

> or something else? Again, I’m very much open to hearing that there’s a strong motivation
> for it, and it’s the one thing that has felt frustrating since installing v4b1.

>
> Edit to add: One reason this is jarring is that ⎇b and ⎇f, and likewise ⇧← and ⇧→

well `shift-{right,left}` have a third behavior (move until the next space).
but as mentioned below, that's probably obsolete.
I guess we should at least change those to token movements, to make the system less confusing.

> still move by the old “word boundary” algorithm, while ⎇Delete and ⇧Delete both delete
> the whole “chunk” now.

Nope, `⇧Delete` (aka `shift-delete`) should not delete an entire token. It should only delete the current history search result / autosuggestion from history, and if there's none falls back `backspace` but we can get rid of that, see below.

---

The reason starts with: there was no default binding for token-wise
left/right/backward-kill/forward-kill (all of which have been very useful IME).

I think we can assume it's better to have default bindings for both token-wise and word-wise rather than just one, so our first lemma is that we want to add them somewhere and keep word-wise operations in other default bindings.

Let's start with pretendng macOS doesn't exist because it has a number of issues[^macos].

- I assume we have three options for modifiers to use, `ctrl`, `alt` and `shift`.
- We could use `shift-{left,right}`, however that's weird for a couple of reasons
  1. In Emacs descendants, `shift-{left,right}` usually extends the selection[^shift-extends]
  2. we don't want to use `shift-{backspace,delete}` for token deletion, because:
     2.1. Users expect that `shift-backspace` (in text-insertion modes) behaves as `backspace`
     2.2. `shift-delete` is already bound (it deletes the currently selected history item or autosuggestion from history)[^shift-delete].
     2.3. In spite of 2.2 and 2.3, we do want a way to delete an entire token.
          - In bash, `ctrl-w` deletes a bigword (until the next space), not a word like in fish/zsh. We could do that but that's also breakage and the asymmetry with token movements would be ugly.
- In bash/zsh, `alt-{b,f,d,left,right,backspace}` operate on a word (`alt-delete` doesn't work here)
- Keys like `alt-b` are pretty fundamental Emacs bindings; I reckon that `alt-backspace` does not carry the same strong expectation.
- Unlike other shells, fish supports `ctrl-{backspace,delete}` OOTB
- Before that change, fish has up to three versions for some movements (for example `ctrl-left`, `alt-left`, `alt-b`), that's a lot of redundancy.
- We could use `ctrl-{left,right,backspace,delete}` for token ops but
  - All popular desktop OS but macOS use `ctrl-{left,right,backspace,delete}` to operate on words.
  - Some terminals don't support that yet (not a strong reason though)

So I arrived at `alt-{left,right,backspace,delete}` as best choice.
Take that with a grain of salt because before that I've never used those keys in fish.

There's a tradeoff to be made between consistency and compatibility.
I have preferred to err on the side of consistency.
Breakage in interactive scenarios isn't unequivocally bad because it means the change is discoverable.
I'm sure we can come up with various alternatives that increase compatibility but I was  waiting until anyone was asking for that.

One alternative: simply use `ctrl-{left,right,backspace,delete}` instead.
That is much less breaking since `ctrl-backspace` and `ctrl-backspace` were essentially unbound pre-4.0.
It is still breaking for `alt-{left,right}`.
I'd wager that I'd probably be fine with that but I'll need to try it for while.

Another quick sketch (this one doesn't sound great to me):
1. Keep `alt-{left,right,backspace,delete` as before
2. `shift-{left,right}` for token movements
2. `ctrl-W` for deleting a token back
2. `alt-D` for deleting a token forward

Disadvantages:
- keys like `alt-D` might be intercepted by the terminal on Linux
- less consistent
- `shift-{left,right}` is weird, see above

Could also try multiple modifiers `alt-shift-delete` etc. but that seems awkward,
especially when using a 60%-keyboard which requires a modifier for delete.

[^macos]: We're in the phase of collecting opinions from macOS users, since we're not sure what's best on that platform. The issues at hand are:
First, most macOS terminals don't default to "option as alt", so many users can't use `alt` bindings at all. I guess those users are not relevant, unless `option-backspace` sends `alt-backspace` (does it?).
Second, `shift-left` (or was it `ctrl-left`?) are used by the window manager
Third, the default terminal cannot disambiguate ctrl-backspace

[^shift-extends]: fish is different here for reasons that are specific to a shell and also fish's history:
we bind `shift-{left,right}` to bigword (until the next space) movements.
This seems to be mostly obsoleted by token movements.
I would be in favor of removing these bindings, potentially replacing them with selection-extending movements.
I haven't proposed that because we'd need a few other Emacs bindings to make text selections usable by default.

[^shift-delete]: Outside a history search if there is no autosuggestion, `shift-delete` acts as `backspace` but I'm 99% sure that's purely a historical artifact that we can probably remove now.
That statement would be wrong if there are terminals where backspace sends shift-delete, but those terminals would already be broken by the fact that it removes the autosuggestion from history if present.
I'm not sure what was the motivation for 7561075af8 (OS X Mavericks keybinding fix, 2013-09-05), maybe this was just about fixing `shift-backspace` not `shift-delete`?

**Comment by chriskrycho** [comment]
Thank you for the thorough response! I read this and am mulling on it, and your broad strokes motivations make sense… but I am also hit directly by footnote (1), since I am a macOS user, and I use Terminal rather than iTerm most of the time.[^latency]

It is possible the behavior of <kbd>⎇</kbd><kbd>Delete</kbd> I saw in zsh is related to something about zsh on macOS, but I consistently see it in both iTerm and Terminal.

I do recognize that iTerm works with <kbd>Ctrl</kbd><kbd>Delete</kbd> and Terminal does not; it’s surprising (but I agree in practice it is true) that <kbd>Ctrl</kbd><kbd>Delete</kbd> does not work in Terminal. That said, on macOS, <kbd>⎇</kbd><kbd>Delete</kbd> *is* bound in normal text areas to work on the word-by-word basis (it’s *basically* `M-backspace` from Emacs, courtesy of the broad support for Emacs-ish key bindings in `CoreText`), so it’s not just a terminal- or shell-specific issue but a macOS compatibility issue in general.

> [!NOTE]
> None of this is meant as opinions on what you should do _per se_ (though obviously I would prefer <kbd>⎇</kbd><kbd>Delete</kbd> to work the way it did before!) so much as it is an attempt to provide useful info!

[^latency]: I am sensitive to terminal typing latency, and [like some other folks][luu] have found iTerm annoyingly higher on that axis than Terminal, so I mostly use Terminal.

[luu]: https://danluu.com/term-latency/

**Comment by krobelus** [comment]
> and I use Terminal rather than iTerm most of the time

ok but it sounds like you did configure Option to send Escape/Alt. I thought that your persona would prefer `alt-b alt-d` to `alt-backspace` since the former doesn't require stretching your pinky.
Of course that doesn't invalidate your point.

> That said, on macOS, ⎇Delete is bound in normal text areas to work on the word-by-word basis

Yeah, I thought it's cleanest to make it platform-dependent (e.g. swap `ctrl-left` and `alt-left` on macOS), so we behave like the rest of the system on both Linux and macOS but there was some push-back at https://github.com/fish-shell/fish-shell/pull/10766#discussion_r1797345099
Using native bindings seems like a progressive, forward-looking approach but maybe that should be done for all keys at once and not just these four.
(We can even try to make client-OS detection work over SSH based on `LC_TERMINAL_PROGRAM` and other environment variables. Containers too)

> None of this is meant as opinions on what you should do per se

I don't have strong opinions on fish's behavior on macOS but I wouldn't want to make the behavior on other platforms significantly worse because macOS is different.

**Comment by chriskrycho** [comment]
> Yeah, I thought it's cleanest to make it platform-dependent (e.g. swap ctrl-left and alt-left on macOS), so we behave like the rest of the system on both Linux and macOS but there was some push-back at https://github.com/fish-shell/fish-shell/pull/10766#discussion_r1797345099
> Using native bindings seems like a progressive, forward-looking approach but maybe that should be done for all keys at once and not just these four.
> (We can even try to make client-OS detection work over SSH based on LC_TERMINAL_PROGRAM and other environment variables. Containers too)

I like this idea. I can also work around it using `bind` in the meantime, but it continues to be a point of *significant* friction for me with the beta.

> I don't have strong opinions on fish's behavior on macOS but I wouldn't want to make the behavior on other platforms significantly worse because macOS is different.

I definitely agree with this, but/and I think there’s a huge tension in terms of ttwo very different sets of users:

- folks who are Mac users first (or even *only*!) who use fish
- folks who are fish users first, who use some mix of macOS, Linux, and other platforms

I sympathize deeply with the challenge here because it is not at all obvious how to get things “right” for both of those groups. I am definitely in the first of those two buckets, but that doesn’t make my needs more important! I wonder if it would be possible to ship sets of bindings that respect those defaults, kind of the way some editors (e.g. Zed) have started shipping keybindings you can opt into from other editors. 🤔 Seems like a non-trivial amount of initial work, but probably *relatively* low maintenance, but of course it’s easy for me to say that from the outside. 😂

**Comment by krobelus** [comment]
I personally think making it native to the current platform is more important (it's also what a web browser would do).
But I also don't know much about macOS. Maybe people who happen to see this can vote here:
1. :+1:  leave things as they are
2. :tada: on macOS, swap ctrl- and alt- left/right/delete/backspace bindings
3. :confused: go back to previous behavior (everything operates on words)

i don't think voting is a good way to go but I don't know how else.
(OT: Great post on mega merges BTW :)

**Comment by chriskrycho** [comment]
I’ve flagged this up for my audience, such as it is, on [Bluesky](https://bsky.app/profile/chriskrycho.com/post/3lf3farbols2b) and [Mastodon](https://mastodon.social/@chriskrycho/113781982210296075), so hopefully more folks will give input here!

(OT: Huzzah!)

**Comment by krobelus** [comment]
Here's my attempt: https://github.com/fish-shell/fish-shell/pull/11107
If we stick to this we can make it work on any other terminals that people may run on macOS (though our workarounds for `Terminal.app` will not work in all scenarios, and it will not work with some terminal multiplexers that don't allow us to talk to the underlying terminal - tmux is fine in this regard).
Not sure what terminals people use. I guess iTerm2, kitty, ghostty, Contour

**Comment by glebpom** [comment]
@krobelus please don't forget about WezTerm

**Comment by faho** [comment]
We reverted the bindings for 4.0 in 378f452eaa0abbe1a6e3d41963adfcee11803a09.

The alternative would be the first time we're gating bindings on OS, which is weird at this point in the release.

I'm going to close this now for 4.0, and then we'll see what we end up with for 4.1.

**Comment by chriskrycho** [comment]
Thanks for the update! Please feel free to ping me (and I’ll also try to keep my eyes open) for future iterations. I think the goals here were and are still good, just a matter of getting the details just right!

**Comment by chapmanjacobd** [comment]
If anyone is looking for the 4.0b behavior just add this to your `config.fish`:

```
bind alt-right nextd-or-forward-token
bind alt-left prevd-or-backward-token

bind alt-backspace backward-kill-token
bind alt-delete kill-token
```

I think the decisions here make sense... I was pleasantly surprised that it was made default in 4.0b though I completely understand how difficult it is to re-train muscle memory. 
The 4.0-final default makes sense for the common case: existing fish users who are used to v3 key bindings as well as new fish users who are used to zsh/bash behavior.

From my perspective (on Linux) the setting above unifies some of the behavior in the browser and terminal (`Ctrl-{Left,Right,Delete,Backspace}` allows moving and deleting words in both). `Alt-{Left,Right,Delete,Backspace}` does nothing in the browser.

It looks like in a browser (Chrome) on Mac OS, `Ctrl-{Left,Right,Delete,Backspace}` does nothing, `Option-{Left,Right,Delete,Backspace}` operates on words and `Command-{Left,Right,Delete,Backspace}` operates on lines (similar to the `Home` and `End` keys).

**Comment by krobelus** [comment]
Right. I'm still going for making it platform-dependent in future. It's a very useful feature.

> Command-{Left,Right,Delete,Backspace} operates on lines (similar to the Home and End keys).

yeah we can add those bindings fairly easily. Could theoretically add them on all platforms though they probably only make sense on macOS.

**Comment by chapmanjacobd** [comment]
Thanks again Johannes for all the help to bring this feature to life

> re: operating on lines

`Shift-{Left,Right}` is similar behavior and already working. Maybe it would be useful to add a `Shift-{Delete,Backspace}` as well... though shift-delete is mapped to backspace in many desktop environments.

I guess my main point in my comment above is that there are very real differences between platforms and different users have different expectations. In my limited experience:

- Mac OS `Option` is similar to Linux `Ctrl`: both move words in the browser
- Mac OS `Ctrl` is similar to Linux `Alt`: both do nothing in the browser

Both Mac OS `Ctrl` and `Command` _could_ be candidates for token operations but Mac OS `Option` should definitely be for word operations. I guess things are already in motion here: https://github.com/fish-shell/fish-shell/pull/11107

**Comment by krobelus** [comment]
added basic platform-dependent bindings in 2bb5cbc9

I'd be happy to make `command-{left,right}` operate on tokens (since `ctrl-{left,right}` is not available) but probably not too many macOS terminals will support it, since iTerm maps them to something else and Ghostty doesn't allow mapping it yet (they copy Terminal.app behavior, even when we turn on the kitty keyboard protocol)

**Comment by rien333** [comment]
The behavior of option+backspace seems to have changed between 4.0.8 and 4.1. Previously, it deleted words, now it basically deletes until a space boundary. May just be a temporary regression. 

> Keys like alt-b are pretty fundamental Emacs bindings; I reckon that alt-backspace does not carry the same strong expectation. 

As an almost 10 year long emacs user, it does. The thing is that these are readline/readline-ish conventions — in any case, they go far beyond emacs and the shell. I haven't used macOS in a while, but back in the day, you had readline shortcuts everywhere (even in safari etc.). Gnome on linux has something pretty similar, though it only works in GTK3 apps. The behavior of fish now no longer matches all these apps. 

And then there is CLIs: on my system, `python` interprets alt-backspace as "delete single word".

**Comment by injust** [comment]
> The behavior of option+backspace seems to have changed between 4.0.8 and 4.1. Previously, it deleted words, now it basically deletes until a space boundary. May just be a temporary regression.

That's https://github.com/fish-shell/fish-shell/commit/2bb5cbc95943b0168c8ceb5b639391299e767e72. It's under https://fishshell.com/docs/current/relnotes.html#new-or-improved-bindings in the 4.1.0 changelog.

IMO it's incredibly annoying that fish keybinds are platform-dependent. When connecting from a macOS system to a Linux box, this is now the first thing I have to fix (after installing fish, that is).

**Comment by krobelus** [comment]
> As an almost 10 year long emacs user, it does. The thing is that these are
> readline/readline-ish conventions — in any case, they go far beyond emacs
> and the shell.

I see - I mainly used Vi key bindings when I was using Emacs.
But `ctrl-backspace` should work fine in many places.

> I haven't used macOS in a while, but back in the day, you
> had readline shortcuts everywhere (even in safari etc.). Gnome on linux has
> something pretty similar, though it only works in GTK3 apps. The behavior
> of fish now no longer matches all these apps.
> And then there is CLIs: on my system, python interprets alt-backspace as
> "delete single word".

@rien333 the way fish 4.1 works is that
1. `ctrl-backspace` work like they do in GUI apps like firefox or chrome.
   The main reason why `ctrl-backspace` does not work OOTB in `emacs -nw` or Python
   is that legacy terminals encode `ctrl-backspace` the same as `ctrl-h`,
   so until those apps use something like the kitty keyboard protocol,
   they cannot distinguish between those two key presses.
   For `emacs -nw` users, there is [kkp.el](https://github.com/benotn/kkp).
2. The key chords that are not used in GUIs (e.g. `alt-backspace` on Linux)
   are used by fish to support a common fish-flavored use case.
   I realize that the change can be annoying,
   but I thought there are plenty alternatives like `ctrl-backspace`, `ctrl-w` and `alt-b,alt-d`
   The latter two already work in readline.

It's not an easy tradeoff, and I'm not sure if it's easy to get enough feedback.

> That's 2bb5cbc. It's under
> https://fishshell.com/docs/current/relnotes.html#new-or-improved-bindings
> in the 4.1.0 changelog.
>
> IMO it's incredibly annoying that fish keybinds are platform-dependent. When
> connecting from a macOS system to a Linux box, this is now the first thing
> I have to fix (after installing fish, that is).

@injust Yes, I definitely recognize that this is annoying, hopefully we can fix this:

Right now we use `$(uname)` in fish but I think should ask the terminal on which OS it runs on.
If you run kitty on macOS connecting to a remote fish on Linux,
then we could make it use macOS key bindings, since kitty already advertises which OS it's running on.

Would that help your use case?
Note that:
- if we go with this approach, we still need ask other terminals to implement this.
- this won't work if fish runs inside `tmux` unless we add another hack
- I guess for now we could add an iTerm2 specific hack (since it's macOS-only)

I think that making it platform dependent is them more future-proof approach -- assuming we want to make fish to behave more like native apps (the browser is the role model).
It does bring birth-pangs since AFAICT we're the first terminal app to try to do this.. but if a GUI app can do it why shouldn't a terminal one?

**Comment by injust** [comment]
> Right now we use `$(uname)` in fish but I think should ask the terminal on which OS it runs on. If you run kitty on macOS connecting to a remote fish on Linux, then we could make it use macOS key bindings, since kitty already advertises which OS it's running on.
> 
> Would that help your use case? Note that:
> 
> * if we go with this approach, we still need ask other terminals to implement this.
> * this won't work if fish runs inside `tmux` unless we add another hack
> * I guess for now we could add an iTerm2 specific hack (since it's macOS-only)
> 
> I think that making it platform dependent is them more future-proof approach -- assuming we want to make fish to behave more like native apps (the browser is the role model). It does bring birth-pangs since AFAICT we're the first terminal app to try to do this.. but if a GUI app can do it why shouldn't a terminal one?

I personally don't use kitty, but this sounds like an interesting approach.

In the meantime, I've simply added the macOS branch of the platform-dependent keybinds to my dotfiles, which works decently as a stopgap fix:

https://github.com/fish-shell/fish-shell/blob/4d67ca7c58a9c4889b561ad54c67b15d808e0659/share/functions/fish_default_key_bindings.fish#L54-L58

**Comment by krobelus** [comment]
> I personally don't use kitty, but this sounds like an interesting approach.

Yes, we've been going back and forth on whether this would become a real (cross-terminal) thing.
We'd only need buy-in from terminals that are typically used on macOS, my list is currently
alacritty, iTerm2, kitty, ghostty, Contour, WezTerm.
(the stock Terminal.app hasn't been reacting to bug reports at all).
I wonder if I'm missing another reasonably popular terminal.

But the fact that client OS detection can't ever work nicely in tmux
is a bummer, but I guess it shouldn't hold us back forever, because
there are alternatives like shpool that don't have this problem.

> In the meantime, I've simply added the macOS branch of the platform-dependent keybinds to my dotfiles, which works decently as a stopgap fix:
> 
> https://github.com/fish-shell/fish-shell/blob/4d67ca7c58a9c4889b561ad54c67b15d808e0659/share/functions/fish_default_key_bindings.fish#L54-L58

Right. You might be missing the {ctrl,alt}-{left,right} keys
(those are in a different file because they are shared with Vi mode, because Vim doesn't support things like {ctrl,alt}-backspace).

I edited the OP with the complete set of binds that were changed,

**Comment by raphael-costa** [comment]
I apologize for the negative feedback but this was a TERRIBLE decision.

Fish wasn't written yesterday, users have a **DECADE** of muscle memory and this change just BROKE IT. It also makes fish keybindings needlessly different from other shells.




For reference, users can fix this terrible issue by creating the file ~/.config/fish/conf.d/old_keybindings.fish with the contents:


    bind alt-backspace backward-kill-word
    bind ctrl-alt-h backward-kill-word
    bind ctrl-backspace backward-kill-token
    bind alt-delete kill-word
    bind ctrl-delete kill-token
